### DISCUSSION TITLE
Properly hiding an alternate encoding of a ORM Model Key

### DISCUSSION THREAD

**Discussion by Xevion** [pydantic/pydantic#4479] [Question]
Hi, I'm using Pydantic with FastAPI and SQLAlchemy for my project.

Right now, most of my models use 64-bit signed integers for their primary keys. I encode them to base64 for the API, and the client is intended to never use the integer - only the base64 ASCII key - for all operations. This is fine for routes as there is no special relationship needed for Pydantic. Unfortunately though, when I wanted to make the Pydantic schema for this, I ran into trouble.

I wanted to have two columns related to the primary key for my model: `id` (type *biginteger*), and `id_ascii` (type _varchar(11)_). For the Pydantic model, I wanted it to only show `id_ascii`, and I wanted it to show up in the model as `id`.

I tested various things to try and get this to work

- Naming the property on the Pydantic model `id` or `ascii_id` with the alias set to the opposite (`Field(None, alias=...)`).
- Using the `alias` property, as said now.
- Using `allow_population_by_field_name` in the Model config.

I was able to get this working, but not without making a 'sacrifice' on the database modeling part.
- I renamed `id` in the code to `id_` and specified the column names to match as they were before. Below is the SQLAlchemy code.

```python
class Upload(Base):
    __tablename__ = "upload"

    id_ = sa.Column("id", sa.BigInteger, primary_key=True, autoincrement=False)
    id = sa.Column("id_ascii", sa.String(11), nullable=False)
```
As you'll see below, I got what I want in effect (response is limited.

```json
{
    "id": "6OIlJzSZl9E",
    "creation_timestamp": "2022-09-05T07:49:32.576600",
    "expiry_timestamp": "2022-09-05T08:49:32",
    "key": "UJotnikGZx4-6OIlJzSZl9E.jpeg"
}
```

If you're wondering what I'm looking for, I'm interested in a solution that only requires changes to the Pydantic model. I don't want to interfere with the SQLAlchemy model, where doing something like `id_` is rather confusing to me.

**Comment by samuelcolvin** [comment]
Use a root validator.

**Reply by Xevion** [reply]
Could you expand on what you mean? I don't understand what you propose I do with a root validator. Validation of the model has nothing to do what I want here. I want to 'rename' or extract the property from a different name here.

**Reply by Xevion** [reply]
@samuelcolvin I'm still pretty interested in what you meant here, could you clarify?

**Reply by samuelcolvin** [reply]
use a root validator to change the name of the relevant field before the rest of validation

E.g. something like

```py
class MyModel(BaseModel):
    ...

    @root_validator(pre=True)
    def switch_name(cls, values):
        try::
            values['name_after'] = values.pip('name_before')
        except KeyError:
            pass
        return values
```

**Reply by Xevion** [reply]
Thanks for the reply. Unfortunately, I was not able to get what you had working, and I did poke around the cause of the issue: GetterDict. it does not allow assignment of any sort.

The code example below is a bit different from the original question, but it bears the same modification requirements more or less.
I want to transform the value of the `user_id` property before validation with an `asciify_identifier` method.

```python
UInt64 = conint(ge=0, lt=2 ** 64)

class UploadRead(UploadBase):
    id: str
    user_id: UInt64

    creation_timestamp: datetime
    expiry_timestamp: datetime

    upload_url: constr(max_length=512)
    cloudflare_id: constr(max_length=1024)
    draft: bool

    @root_validator(pre=True)
    def switch_name(cls, values: GetterDict):
        try:
            values['user_id'] = asciify_identifier(values['user_id'])
        except ValueError:
            pass
        return values
```

The runtime below occurs when I use the `UploadRead` model.

```python
  File "F:\Programming\Projects\motophoto\server\.\api\endpoints\upload.py", line 62, in create_upload
    return UploadRead.from_orm(new_upload)
  File "pydantic\main.py", line 579, in pydantic.main.BaseModel.from_orm
pydantic.error_wrappers.ValidationError: 1 validation error for UploadRead
__root__
  'GetterDict' object does not support item assignment (type=type_error)
  ```

The issue I am running into bears most resemblance to issue #821. Getting this working isn't an imminent priority, so I'd like to hear what you think.

**Reply by Xevion** [reply]
@samuelcolvin Sorry to bother you, but this is still an issue I am interested in developing a solution for. Could you take a look at what I have so far and let me know your thoughts? Thanks!
